Source/AwsCommonRuntimeKit/auth/imds/CRTIMDSClient.swift

```diff
@@ -2,28 +2,33 @@
 //  SPDX-License-Identifier: Apache-2.0.
 
 import AwsCAuth
-
+// TODO: fix pointers and look for other errors
 public class CRTIMDSClient {
     let rawValue: OpaquePointer
-
-    public init(options: CRTIMDSClientOptions, allocator: Allocator = defaultAllocator) {
+    let allocator: Allocator
+    public init(options: CRTIMDSClientOptions, allocator: Allocator = defaultAllocator) throws {
+        self.allocator = allocator
         let shutDownOptions = CRTIMDSClient.setUpShutDownOptions(shutDownOptions: options.shutDownOptions)
         var imdsOptions = aws_imds_client_options(shutdown_options: shutDownOptions,
                                                   bootstrap: options.bootstrap.rawValue,
                                                   retry_strategy: options.retryStrategy.rawValue,
                                                   imds_version: options.protocolVersion.rawValue,
                                                   function_table: nil)
-        self.rawValue = aws_imds_client_new(allocator.rawValue, &imdsOptions)
+        guard let rawValue = aws_imds_client_new(allocator.rawValue, &imdsOptions) else {
+            throw CommonRunTimeError.crtError(CRTError.makeFromLastError())
+        }
+        self.rawValue = rawValue
     }
     /// Queries a generic resource (string) from the ec2 instance metadata document
     ///
     /// - Parameters:
     ///    - resourcePath: `String` path of the resource to query
     public func getResource(resourcePath: String) async throws -> String? {
+        let resourcePathCursor = AWSStringByteCursor(resourcePath)
         return try await withCheckedThrowingContinuation { (continuation: ResourceContinuation) in
             let callbackData = CRTIMDSClientResourceCallbackData(continuation: continuation)
             let pointer: UnsafeMutableRawPointer = fromPointer(ptr: callbackData)
-            aws_imds_client_get_resource_async(rawValue, resourcePath.awsByteCursor, resourceCallback, pointer)
+            aws_imds_client_get_resource_async(rawValue, resourcePathCursor.byteCursor, resourceCallback, pointer)
```

Comment:
The pointer is invalid as soon as we leave the closure. So that doesn’t work in async logic where we need a manually managed byte cursor. We can not make a copy because that copy will also be invalid. Here, in particular, I think we can use a closure because of the Continuation which will wait until the callback has fired. However, [SigningConfig](https://github.com/awslabs/aws-crt-swift/blob/b4c6f28a32568c101fa7dc801d5d223b935b7fd2/Source/AwsCommonRuntimeKit/auth/signing/SigningConfig.swift ) is a good example where we need a persistent byte cursor and we cannot use a closure. 

That’s where I ran into the issue of undefined behavior because byte cursor was stored by C struct which was not valid after the init call. We probably can get rid of it after some major refactoring of the signing logic and the way code is organized. I will try to do that in the Auth layer refactoring PR.

Reply:
HttpProxyOptions have a similar issue.

Reply:
Thanks, will resolve in a new PR.